### ISSUE TITLE
allow encoding and decoding of values larger than 32 bits

### ISSUE THREAD

**Issue by jbathegit** [NOAA-EMC/NCEPLIBS-bufr#195]
It's possible for the library to be asked to encode and decode values which occupy more than 32 bits.  An example of this is the WMO standard Table D BUFR sequence 3-15-010, where an operator is used to encode the constituent Table B BUFR descriptor 0-22-069 (spectral wave density) with 38 bits.

Currently, the only way to encode and decode such a value is by using the _8 build build of the library, which defaults to 8-bytes for **all** declared integers.  However, since a long-term goal is to consolidate all of the various _4, _8, and _d builds down to one (and, ideally, to the equivalent of the current _d build - see #78), then we need to figure out a way to encode values >32 bits in **any** of the current builds, by drilling down into the Fortran code to figure out what internal variables need to be explicitly declared as 8-byte integers, while continuing to allow everything else in the code to be declared as 4-byte integers.  This will allow for maximum library efficacy and portability in the future.

**Comment by jbathegit** [comment]
After further investigation, the main internal variable that would need to be declared as an 8-byte integer, for both encoding and decoding, is array IVAL within moda_ival.F.  However, we'd then need to correspondingly declare the following additional internal variables as 8-byte integers as well, since they interact directly with IVAL:

- NVAL and INT within upbb.f
- statement function MPS within rdtree.f
- NVAL, (local) IVAL, and INT within pkb.f
- IPKS within ipks.f

and there will likely be others as well if we keep digging even further.

So, as it turns out, the safest long-term approach will be to **not** change any source code, but rather to just plan to migrate all of the various builds down to the equivalent of the current _8 build instead of the _d build.  That way, everything will be guaranteed to continue to work fine, just as it already does now for the test_OUT_6.f test case with Table B BUFR descriptor 0-22-069 using the current _8 build.

**Comment by jbathegit** [comment]
I am re-opening this issue.  After working further on #78, it's become clear that we'll be better off in the long run if we go ahead and migrate all of the various builds down to _4 or _d after all, rather than _8.

As noted in the discussion for #78, most application codes are likely already coded with 4-byte integers and linking to either the current _4 or _d build, so if we build the consolidated library with 8-byte integers as _8, then we'll end up having to do a lot of extra copying from 4-byte arrays to 8-byte arrays inside of the library interface at run time, which in turn will lead to slower performance.  Furthermore, 4-byte Fortran integers usually better align with the c_int type that is passed in from the C functions defined within the bufr_interface.f90 module, which in turn means there would be less copying of arrays when any of these same library routines are called from within C application codes.

So, bottom line, it will be better to bite the bullet now and work towards a consolidated library build that corresponds to the compiled integer size in the majority of application codes.  I'll go ahead and make whatever changes are needed to the library so that only the handful of internal integer variables which really need to be 8-bytes are explicitly declared as such.

As part of this update, I'll also expand the current CI test suite, so that going forward **all** IN and OUT test application codes will be compiled and tested using each of the _4, _8 and _d variations.  The reason is because several of those codes now explicitly require linking with the _8 build of the library in order to test successfully, so if I can get those same codes to now test successfully with the _4 and _d builds of the library as well, then that will prove that the changes for this update work as intended.  Furthermore, it will set us up nicely for future work on #78, since we'll already have the framework in place to build all of the IN and OUT test application codes using all of the _4, _8, and _d variations, and that way we can verify that all of those variations of application codes continue to work as we progress in the effort to consolidate the library itself down to one build.

**Comment by jbathegit** [comment]
Well, that didn't take long for me to hit a potential snag ;-)

Currently, subroutine PKB will abort if we try to pack an integer larger than 32 bits into a array of 4-byte integers.  In such cases, NBITS will be larger than NBITW, so we can't pack an integer larger than 32 bits in the _4 or _d variations of the library.

@jack-woollen is there any way that subroutine PKB can be modified to allow this?  Otherwise, I don't see any other option than going back to our previous idea of migrating all of the various builds down to _8, and making that the new consolidated version.

**Comment by jack-woollen** [comment]
How about a pass through in pkb to handle the 8 byte case. So instead of jumping to 900 when nbits > nbitw, call pkb8 instead. Also need irev8. Something similar for upbb. Maybe a few others.

**Comment by jack-woollen** [comment]
Another way. If nbits>32, enter an if block which calls pkb recursively, twice, once for each of the two 4 byte words contained in 8 byte nval, high order part first, using the appropriate nbits for each part. Then return.

**Comment by jbathegit** [comment]
Thanks Jack.  Off the top of my head, the second option sounds better to me, because that way it's less additional subroutines and functions we have to document and maintain going forward.  I think it would be better in the long run if everything can be handled directly inside of pkb, using recursion where needed.

I'll wait to hear back from you before proceeding any further on this.  Just an FYI that I'll be away on travel to NAVO next week from Monday thru Thursday, in case you don't hear from me :-)

**Comment by jack-woollen** [comment]
Okay, I'll work up a prototype for option 2 in the mean time. Have a good trip!

**Comment by jack-woollen** [comment]
@jbathegit I have a prototype one build system which passes all the current tests. We need to make some more tests and try to break it! Its now sitting in the same fork as the PR came from. It took three new subroutines and a handful (or two) of other changes. I was thinking of putting it in a branch off develop from the NOAA-EMC/NCEPLIBS-bufr repository to work with it. Maybe replace the one build branch. What do you think?

**Comment by jbathegit** [comment]
Thanks @jack-woollen, and I'd definitely like to take a look at what you've come up with, though I'm not sure what/where you meant by "in the same fork as the PR came from", so I wasn't sure where to look.

For now, could you please put your changes in a new branch off of the develop branch?  I'd rather leave the existing onebuild branch as is until we're sure we won't need it any more.

**Comment by jack-woollen** [comment]
@jbathegit  Made branch jw1build with the changes. How do I run a test build on for this on github?

**Comment by jbathegit** [comment]
Thanks Jack for all of your work on this!  I'm sorry I couldn't get back to you sooner, but this has been a crazy week for me. 
 Overall, your `jw1build` branch looks pretty good, but I do have a few questions as noted below.

First of all, and as an aside, I think you and I may have slightly different ideas of what we mean when we say "onebuild" or "1build" ;-)  In my case, I was using it to refer to the broader issue of consolidating all of the current _4, _8, and _d builds down to a single new build, as outlined in #78.  That will be a longer-term effort, and this particular issue is just a piece of that broader effort, albeit a very important and necessary one to ensure that values larger than 32 bits can be encoded and decoded properly within the eventual unified build.  So, again, I definitely appreciate your efforts here!

Secondly, and to answer your earlier question, the way you can automatically test any new branch containing changes associated with an open issue in GitHub is to open a new PR for your branch, then link it to the issue in question by including the following text as the first line in a discussion block:

`Fixes #X`

if it completely resolves issue #X, or else:

`Part of #X`

if it partially resolves issue #X.  Either way, once you submit the PR, your changes will automatically be tested within the GitHub runner across all supported test environments, and you'll be able to see the results right there inside of the PR.  Furthermore, if you need to later amend your PR with additional commits, you can do so by pushing those commits onto the end of the same branch, and the corresponding PR will automatically re-test your full set of changes.

Now, for my specific questions re: your `jw1build` branch:

- I presume your intent is to have up8 be an 8-byte variant of upb, upb8 be an 8-byte variant of upbb, and pk8 be an 8-byte variant of pkb, correct?  If so, then that's fine, but I think they should each be in their own source file, instead of combined into a single atepak.f source file.  We'll also need some documentation for each new routine, at minimum a descriptive docblock.
- On a related note, what are the upb8p and bits routines for?  I didn't see anywhere those were being called, unless I missed something.
- Why is a "print *, str" line being added to the bort routine?
- Why are the conditional compilation directives being removed from the irev routine?
- Why is the test for the test_OUT_6 code now commented out (i.e. why do you believe this test is now irrelevant)?  If anything, keeping this test is crucial to ensure that we can still properly encode values larger than 32 bits.  I think a better approach would be to keep this test, but then expand the testing so that **all** IN and OUT test application codes get compiled and tested using each of the _4, _8 and _d library builds going forward.  This is noted in the earlier discussion for this issue, and that way we can ensure that test_OUT_6 now works properly even when linked with the _4 or _d build of the library.  Then, once we eventually fully resolve #78, we can continue to keep all of the separate _4, _8, and _d test cases for all of the IN and OUT test application codes, but then at that point we just simply link all of those test cases to the new unified library build, rather than to the separate _4, _8, and _d builds.
- What's the reason for the changes to the binv and readmp utilities?  Unless these are somehow related to this issue, then I think these should be modified using their own separate issue(s) and PR(s), rather than shoehorning them into this one.

**Comment by jack-woollen** [comment]
I'll start with specific questions about enabling >32 bit elements: 

The atepak is just a prototype version of the new subroutines bundled for convenience in debugging. The 'extra' ones helped debugging. I'll dress up pk8, up8, and upb8 properly for a pr and lose the others. 

I added a print statement in bort because the output from testing the branch in the repo evidently doesn't include stderr output where the bort statements are printed. Is there some other way to see stderr messages during repo testing? By the way I found that when I push to the branch it runs the tests anyway, It just doesn't report anything if the tests pass.

At some point I removed the conditional compile statements from irev for debugging. I meant to restore them.

I'll isolate the binv and readmp changes to a separate branch and pr.

The last answer is about removing the out_6_8 test. The new code isn't compatible with the 8byte lib. My thought is why should it be? That test can be modified to call the one build version. This is where I see the connection between this development and the move to a single build. I also think the 4 byte version should become the one build. To begin with, I completely agree with your prior comment below.

_So, bottom line, it will be better to bite the bullet now and work towards a consolidated library build that corresponds to the compiled integer size in the majority of application codes. I'll go ahead and make whatever changes are needed to the library so that only the handful of internal integer variables which really need to be 8-bytes are explicitly declared as such._

That's basically what I think this is.  As you say transition to one build is a process and will take some time. That's exactly why the process should begin. The question is how to proceed. Taking a page from the NCO book for phase-in and phase-out systems, I'd say freeze the 2 deprecated versions and disconnect them from the one build version. This gives developers incentives to move to the one build, along with some time while the frozen versions are still available for production.  Since adapting from 4 or d or 8 byte lib to the one build version is fairly simple, that should be sufficient to get the job done.

**Comment by jbathegit** [comment]
> I added a print statement in bort because the output from testing the branch in the repo evidently doesn't include stderr output where the bort statements are printed. Is there some other way to see stderr messages during repo testing? 


When you're testing locally on a machine using ctest, any bort messages should show up in your build/Testing/Temporary directory, within the file LastTest.log.  They do for me, at least.

You can also manually re-test any of the test codes that were automatically built if you go to the build/test directory and then manually run the corresponding test executable from the command line.  In other words, you don't first need to re-compile anything manually, and if you manually run any test executable from the command line then you should also see any bort output that way as well.




> 
> The last answer is about removing the out_6_8 test. The new code isn't compatible with the 8byte lib. My thought is why should it be? That test can be modified to call the one build version. This is where I see the connection between this development and the move to a single build. I also think the 4 byte version should become the one build. To begin with, I completely agree with your prior comment below.
> 
> _So, bottom line, it will be better to bite the bullet now and work towards a consolidated library build that corresponds to the compiled integer size in the majority of application codes. I'll go ahead and make whatever changes are needed to the library so that only the handful of internal integer variables which really need to be 8-bytes are explicitly declared as such._
> 
> That's basically what I think this is. As you say transition to one build is a process and will take some time. That's exactly why the process should begin. The question is how to proceed. Taking a page from the NCO book for phase-in and phase-out systems, I'd say freeze the 2 deprecated versions and disconnect them from the one build version. This gives developers incentives to move to the one build, along with some time while the frozen versions are still available for production. Since adapting from 4 or d or 8 byte lib to the one build version is fairly simple, that should be sufficient to get the job done.


I understand that you're suggesting to stop building the _d and _8 versions of the library right now, rather than later down the road.  In that case, we'd need to make some corresponding updates to several of the CMakeLists.txt files so that it only builds the equivalent of the current _4 version going forward.  But we'd also need to modify the test/CMakeLists.txt file so that it now builds and tests separate executables for each of the IN and OUT codes using all of the _4, _8, and _d flag variations, and linking each of them with the new unified library build.  That way, we can ensure that the new unified library build continues to work properly with any application codes that may continue to be compiled using any of the _4, _8, and _d flag variations.  Otherwise, how would we be able to confirm that any library updates work correctly for the test_OUT_6 code to encode values larger than 32 bits?

I'm certainly willing to help with this, including any needed updates to the various CMakeLists.txt files to move us down that path, but there's a significant problem I see with going immediately to a unified library build vs. deferring that to later in the process, and which is why I hadn't made that move myself yet.  Specifically, and unless I'm missing something, all of the user-callable routines within the unified library first need to be modified to work automatically no matter whether they're called from an application code compiled using _4, _8, or _d settings.  For example, when an application code calls openbf, the openbf routine needs to be able to work no matter whether the 1st and 3rd arguments being passed in are 4-byte integers or 8-byte integers, but if we're no longer linking the application code to a specific _4 or _8 byte build of the library, then the routine has no way of knowing ahead of time what size those integers are.  That's why in my `onebuild` branch I've been working to develop generic interfaces in Fortran90 for all of the user-callable library routines such as openbf, in order to be able to handle those variations automatically inside of the library.

In short, I don't see how we can move this quickly to a single, unified library build without significantly reducing the amount of CI testing we're doing, and which in my mind could be a recipe for disaster down the road.  Or am I missing something here?

**Comment by jack-woollen** [comment]
Yeah, its a big lift to make the bufrlib routines insensitive to argument kinds. Having looked at it in some detail myself, and seeing what you have been doing in the onebuild branch, I decided its way more trouble than its worth. Any code currently compiled with integer(8) will only have to define integer arguments to unified bufrlib routines as integer(4). Since there are relatively few codes like that (I could only find 2 or 3 in production libraries that link with bufrlib8),  that doesn't seem to be too big of a burden..

**Comment by jbathegit** [comment]
Hey Jack, just touching base to clarify where we are on this.  Maybe I misunderstood, but are you still working on this?

Specifically, are you still working on changes to the library so that values larger than 32 bits can be encoded and decoded for all _4, _8, and _d builds?  As noted earlier, I see that as an important piece of the larger goal to eventually have a single unified build of the library, which I agree should be akin to the current _4 build.

Please let me know - thanks!

**Comment by jack-woollen** [comment]
I guess you're thinking we need a one size fits all in all sizes, maybe to ease to code conversion to one build. Maybe it can be done. I'll look in that direction.

**Comment by jbathegit** [comment]
Yes, as I noted earlier, I think the most sensible first step is to get the library to a state where each of the IN and OUT test application codes can be compiled and tested using each of the _4, _8 and _d variations.

As you know, several of those codes (such as test_OUT_6) now explicitly require linking with the _8 build of the library in order to test successfully.  Since the long-term goal is a single build of the library that works with any _4, _8, or _d application code, then it seems to me we're taking a step backwards if we now turn off any of those existing tests; rather, we need to modify the library to get those same codes to now test successfully with the _4 and _d builds as well, which will prove that any changes associated with this issue work as intended.  In other words, we need a net increase in the number of tests, not a decrease.

Then, once we have a framework in place where each of the test codes works for each of the _4, _8, and _d library builds, that will set us up nicely for future work on https://github.com/NOAA-EMC/NCEPLIBS-bufr/issues/78, because we'll always be able to verify that each of those builds of each application code continues to work as we progress in the effort to consolidate the library itself down to one build.  In other words, we'll continue to have 3 separate tests for each application code (i.e. we'll always compile and test each one with _4, _8, and _d flags), but eventually each of those tests will link to the same unified build of the library, rather than each one to its own separate _4, _8 or _d build of the library, as they will initially.  Does that make sense?

Otherwise, if it turns out to not be possible to modify the current library to get this issue resolved, then the only other alternative I see is that we implement all of the generic interfaces to all of the user-callable routines first (i.e. do all of the rest of the work for #78 first, to get us much closer to having a unified build of the library).  Then, at that point, we could revisit this issue just within that one build, and then something akin to what you have now may well work.

Thoughts?  Please keep me posted - thanks!

**Comment by jack-woollen** [comment]
The jw1build branch now passes all tests, and works with all size builds.

**Comment by jack-woollen** [comment]
@jbathegit New branch jw64ints was started from develop, has just the encode/decode stuff from jw1build, and all subroutines separated and doc blocked. It passes all the tests including test_OUT_6_8. I was thinking of adding a test to see if it actually will store an element >32 bits. Any particulars you want to see before I make a PR?

**Comment by jbathegit** [comment]
Thanks Jack!

At the moment, I'm focusing on getting all of the recent commits to the `develop` branch bundled up into a new 11.7.0 release, as requested by Ron McLaren in #206.  As soon as that's done, I'll take a deep dive into your latest branch and then be ready to discuss moving forward with that as part of the next future release.  Just give me another day or two ;-)

**Comment by jbathegit** [comment]
Thanks again Jack for your efforts on this!  The 11.7.0 release is now done, and I've managed to clear some other tasks off my docket, so I'll be taking some time today to look over your new `jw64ints` branch.

A couple of initial thoughts/comments:
1. test_OUT_6 already tests the storing (encoding) of an element > 32 bits.  I think what we really need is a new test code which reads (decodes) such an element, in order to ensure that works too.
2. On a related note, and as I mentioned earlier, I think we also need to expand the testing so that **_all_** of the IN and OUT test codes are now built with **_all_** of the _4, _8, _d variations.  That will really prove that the new `jw64ints` branch works as intended, if we can now successfully run, e.g. test_OUT_6 when it's built as _4.  Otherwise, we aren't testing any new combinations that weren't already working before with the previous version of the code.  Furthermore, we're going to want all of those tests in the future anyway, even once we migrate the library itself down to one unified build, because we'll always want to ensure that **_all_** application codes (no matter whether they're built as _4, _8, or _d) continue to work with the one unified build of the library.

**Comment by jack-woollen** [comment]
Jeff, sure thing. I agree with all that. I did move in that direction by expanding the out_6 test to both 4 and 8 byte builds, and it passes both. And, the more tests the better of course. I'll expand the other tests too and see what happens.

**Comment by jack-woollen** [comment]
FWIW, jw64ints passed all 4 and 8 byte IN and OUT tests.

**Comment by jbathegit** [comment]
Awesome Jack!  To seal the deal, could you maybe try adding _d tests as well?  That would cover all the proverbial bases.

In the meantime, a couple of initial questions from looking over your `jwints64` branch:
1. In ipks.f, why are you adding .5 to the calculation for the IPKS variable?  Is this to force some sort of rounding behavior?  Instead, why not just replace the use of NINT with IDNINT, which is designed for use with 8-byte computations?
2. In each of ufbget.f, ufbtab.f, and ufbtam.f, I see you have a statement function "int(int8) = int8" declared, which you're then using as a wrapper for the 3rd argument in a call to usrtpl.   Instead, why not just use the same solution you used in rdcmps.f, where you just stored the 8-byte value inside a temporary 4-byte NBMP variable and then passed NBMP into usrtpl?  That way, you don't have to declare that statement function in each of those routines.  I'm actually surprised that even worked and didn't cause the compiler to squawk, given that int() is a Fortran intrinsic, and so I would have thought that was already a reserved function name(?)

**Comment by jack-woollen** [comment]
1> adding .5 always does the right thing. NINT, IDNINT, and other similar intrinsic functions do not have standard behavior across compilers regarding word sizes.
2> six of one half dozen of the other. Note this works across all the int sizes and compilers in the cmake tests.

**Comment by jbathegit** [comment]
> 2> six of one half dozen of the other. Note this works across all the int sizes and compilers in the cmake tests.

Fair enough, but I think we should be consistent and not use different approaches in different routines, which may only lead to confusion down the road.  I personally vote for the NBMP approach as used in rdcmps.f, because I think that's cleaner and clearer.

**Comment by jack-woollen** [comment]
Ok, I go with simple. So changed that, and added the d tests for the in/out tests, and all passed on orion where I;m working today. But when I pushed up to the remote branch a number of the tests failed to compile in github. I don't understand the errors it is having, For example below. More in the actions window outputs.

/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/test/test_IN_1.f:122:18:

  122 |      +   ( IDNINT(r8arr(3,1)) .eq. 14 ) .and.
      |                  1
Error: ‘a’ argument of ‘idnint’ intrinsic at (1) must be double precision
/home/runner/work/NCEPLIBS-bufr/NCEPLIBS-bufr/bufr/test/test_IN_1.f:138:18:

  138 |      +   ( IDNINT(r8arr(1,186)) .eq. 2523 ) ) THEN
      |                  1
Error: ‘a’ argument of ‘idnint’ intrinsic at (1) must be double precision
make[2]: *** [test/CMakeFiles/test_IN_1_8.x.dir/build.make:75: test/CMakeFiles/test_IN_1_8.x.dir/test_IN_1.f.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:807: test/CMakeFiles/test_IN_1_8.x.dir/all] Error 2

**Comment by jbathegit** [comment]
Hmm, that's weird.  I'll take a look.

**Comment by jbathegit** [comment]
Jack, what compiler were you using on orion?  Just curious b/c I don't have any problems with the Intel compiler on WCOSS2.  The error seems to show up with the gfortan9 (gnu) compiler, so again just curious what compiler you were using which worked fine on orion?

**Comment by jack-woollen** [comment]
Intel 2020.2 on orion. What I thought was weird is the error makes no sense. The argument to idnint is real(8).

**Comment by jbathegit** [comment]
Yeah, I agree this is a head-scratcher.  I'll dig some more.

**Comment by jbathegit** [comment]
I was able to reproduce the error on WCOSS2 using the gcc/11.2.0 (gnu) module.  I researched and played around with some different suggested options, and eventually I was able to resolve it by changing all of the IDNINT references to NINT, which apparently is now a more generic option for multiple input types, including complex variables.  This change also still works for Intel (I did check ;-)

I still have no idea why the gnu compiler mistakenly thought we were passing IDNINT something that wasn't double precision, but hey, whatever!?

At any rate, I'll push up another commit on top of this branch later today to resolve this.  In the meantime, I have some other unrelated tasks I need to attend to, and I also want to add a new IN test for decoding values larger than 32 bits.  I'll let you know once that's done, and then at that point we can hopefully get this branch merged and put this issue to bed! ;-)

**Comment by jack-woollen** [comment]
Great! Which branch will you push the modified tests into?

**Comment by jbathegit** [comment]
I can plan to just push everything up on top of your existing `jw64ints` branch, if that's OK with you?

Some of this is ready to go, but I just haven't finished it all up yet b/c I keep getting sidetracked by other issues related to the WCOSS2 30-day test.  I'll keep you posted...

**Comment by jack-woollen** [comment]
Yup, that's fine.

**Comment by jbathegit** [comment]
I just pushed up a commit which includes everything except the new IN test, which I just ran out of time/steam to finish up today after dealing with a bunch of other stuff.  But I figured this way you could at least look over my other updates in the meantime and let me know if you have any questions or concerns.  I'll try to get the new IN test added on Tuesday, and then we can be done with this issue.

Have a great weekend Jack, and thanks again for all your work and other assistance to help get this done!

**Comment by jack-woollen** [comment]
Looks good. I reviewed the changes and only two I don't understand.

diff --git a/src/up8.f b/src/up8.f
-C> USAGE:    CALL UP8 (NVAL, NBITS, IBAY, IBIT)
+C> USAGE:    CALL UPB (NVAL, NBITS, IBAY, IBIT)

diff --git a/src/upb8.f b/src/upb8.f
-C> USAGE:    CALL UPB8 (NVAL, NBITS, IBIT, IBAY)
+C> USAGE:    CALL UPBB (NVAL, NBITS, IBIT, IBAY)

**Comment by jbathegit** [comment]
I agree the USAGE statements should reference UP8 and UPB8, rather than UPB and UPBB.  But they already are referencing the correct names, at least as I see it in the latest commit and in my clone copy on WCOSS2.  Are you sure you're displaying your diff output in the correct order (previous vs. latest, rather than latest vs. previous)?

Or am I missing something here?

**Comment by jack-woollen** [comment]
Actually, that's right, I was looking at the last two commits in the wrong order. Sorry.

**Comment by jbathegit** [comment]
No worries.

BTW, I'm working on the new test code as we speak.  It will have a test for reading values larger than 32 bits, but I'm also going to add in some additional tests for other subprograms that aren't already being tested in any of the other codes.  I figured I might as well, since I'm already standing up a new test code anyway.  Will keep you posted...

**Comment by jack-woollen** [comment]
When the comparison is reversed your changes are much clearer. Duh. Good idea to have more tests.

**Comment by jack-woollen** [comment]
When the comparison is reversed your changes are much clearer. Duh. Good idea to have more tests.
